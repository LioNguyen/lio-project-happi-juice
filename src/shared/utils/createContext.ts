import type React from "react";
import { createContext as originCreateContext, useContext } from "react";

function createContext<T>(): [React.Provider<T>, () => T, React.Context<T>] {
  const context = originCreateContext<T>({} as T);

  return [context.Provider, () => useContext(context), context];
}

export { createContext };
