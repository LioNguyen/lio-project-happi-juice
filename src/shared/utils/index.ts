export * from './lazyLoad'
export * from './tailwind'
