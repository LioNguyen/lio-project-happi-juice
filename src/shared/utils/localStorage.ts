/**
 * Save data to localStorage
 * @param key - The key under which the data will be stored
 * @param value - The data to store, can be any serializable value
 */
function setLocalStorage<T>(key: string, value: T): void {
  try {
    const serializedValue =
      typeof value === 'string' ? value : JSON.stringify(value)

    localStorage.setItem(key, serializedValue)
  } catch (error) {
    console.error(`Error saving ${key} to localStorage:`, error)
  }
}

/**
 * Retrieve data from localStorage
 * @param key - The key of the stored data
 * @returns The parsed data or null if not found or error occurs
 */
function getLocalStorage<T>(key: string): T | null {
  try {
    const serializedValue = localStorage.getItem(key)

    if (!serializedValue || serializedValue === 'undefined') {
      return null
    }

    // Try to parse JSON
    try {
      return JSON.parse(serializedValue) as T
    } catch {
      // Return as string if JSON parsing fails
      return serializedValue as T
    }
  } catch (error) {
    console.error(`Error reading ${key} from localStorage:`, error)
    return null
  }
}

/**
 * Remove data from localStorage
 * @param key - The key to remove from localStorage
 */
function removeLocalStorage(key: string): void {
  try {
    localStorage.removeItem(key)
  } catch (error) {
    console.error(`Error removing ${key} from localStorage:`, error)
  }
}

/**
 * Clear all data from localStorage
 */
function clearLocalStorage(): void {
  try {
    localStorage.clear()
  } catch (error) {
    console.error('Error clearing localStorage:', error)
  }
}

export {
  clearLocalStorage,
  getLocalStorage,
  removeLocalStorage,
  setLocalStorage,
}
