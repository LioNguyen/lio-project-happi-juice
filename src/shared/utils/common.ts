import { format, isValid, Locale, parseISO } from 'date-fns'
import { enUS, vi } from 'date-fns/locale'
import _ from 'lodash'

import { IOrderItem } from '@/domains/order'
import { ISelectOption } from '@/shared/types'

interface INumberFormatOptions {
  decimals?: number
  thousandSeparator?: string
  decimalSeparator?: string
}

// ==================== Date Utilities ====================
/**
 * Parse various date formats into Date object
 * @category Date
 * @param {Date | string | undefined} value - Input date
 * @returns {Date | undefined} Parsed Date object or undefined if invalid
 * @example
 * parseDate("2024-02-26") // Date object
 * parseDate("invalid") // undefined
 */
function parseDate(value: Date | string | undefined): Date | undefined {
  if (!value) return undefined
  if (value instanceof Date) return isValid(value) ? value : undefined

  try {
    // Try ISO string parsing first
    const date = parseISO(value)
    if (isValid(date)) return date

    // Fallback to direct Date creation
    const normalDate = new Date(value)
    return isValid(normalDate) ? normalDate : undefined
  } catch {
    return undefined
  }
}

/**
 * Format a date string according to specified format
 * @category Date
 * @param {string} date - Date string to format
 * @param {string} [dateFormat] - Output format (defaults to "yyyy-MM-dd")
 * @returns {string} Formatted date string
 * @example
 * formatDateTime("2024-02-26") // "2024-02-26"
 * formatDateTime("2024-02-26", "dd/MM/yyyy") // "26/02/2024"
 */
function formatDateTime(date: string, dateFormat?: string) {
  return format(date, dateFormat || 'yyyy-MM-dd')
}

const localeMap: Record<string, Locale> = {
  vi: vi,
  en: enUS,
}

// ==================== Number & Price Utilities ====================
/**
 * Format a number with customizable separators and decimals
 * @category Number
 * @param {number} value - Number to format
 * @param {INumberFormatOptions} [options] - Formatting options
 * @returns {string} Formatted number string
 */
const formatNumber = (
  value: number,
  options?: INumberFormatOptions,
): string => {
  const {
    decimals = 0,
    thousandSeparator = ',',
    decimalSeparator = '.',
  } = options || {}

  const parts = value.toFixed(decimals).split('.')
  parts[0] = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, thousandSeparator)

  return parts.join(decimalSeparator)
}

/**
 * Formats the price with thousand separators using locale-specific formatting
 * @category Price
 * @param {string} price - Price string to format
 * @returns {string} Formatted price string
 */
const formatPrice = (price: string) => {
  return Number(price).toLocaleString()
}

// ==================== Array Utilities ====================
/**
 * Transforms an array of select options into a sorted, joined string
 * @category Array
 * @param {ISelectOption[]} array - Array of select options
 * @param {string} [key] - Key to extract from each option
 * @param {string} [separator] - Separator for joining values
 * @returns {string} Sorted and joined string
 */
const transformToSortedString = (
  array: ISelectOption[],
  key: string = 'value',
  separator: string = ',',
): string => {
  return _.chain(array).map(key).sortBy().join(separator).value()
}

// ==================== Order Utilities ====================
/**
 * Calculates the total price for an array of order items
 * @category Order
 * @param {IOrderItem[]} items - Array of order items
 * @returns {number} Total price
 */
const calculateTotalPrice = (items: IOrderItem[]) => {
  return items.reduce((sum, item) => sum + item.price * item.quantity, 0)
}

export {
  // Date utils
  parseDate,
  formatDateTime,
  localeMap,

  // Number & Price utils
  formatNumber,
  formatPrice,

  // Array utils
  transformToSortedString,
  // Order utils
  calculateTotalPrice,
}
