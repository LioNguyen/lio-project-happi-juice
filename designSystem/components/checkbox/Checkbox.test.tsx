import { screen, fireEvent } from "@testing-library/react";
import { ComponentPropsWithRef } from "react";
import { describe, expect, it, vi } from "vitest";
import { renderWithTheme } from "@/shared/utils";
import { Checkbox } from "./Checkbox";

// Mock radix-ui checkbox with proper event handling
vi.mock("@radix-ui/react-checkbox", () => ({
  Root: vi.fn(({ children, className, checked, disabled, onCheckedChange, ...props }) => (
    <button
      type="button"
      data-testid="checkbox-root"
      data-state={checked ? "checked" : "unchecked"}
      disabled={disabled}
      className={className}
      onClick={() => onCheckedChange?.(!checked)}
      {...props}
    >
      {children}
    </button>
  )),
  Indicator: vi.fn(({ children, className }) => (
    <span data-testid="checkbox-indicator" className={className}>
      {children}
    </span>
  )),
}));

// Mock lucide-react
vi.mock("lucide-react", () => ({
  Check: () => <div data-testid="check-icon" />,
}));

type CheckboxProps = ComponentPropsWithRef<typeof Checkbox>;

const defaultProps: Partial<CheckboxProps> = {};

const renderComponent = (props: Partial<CheckboxProps> = {}) => {
  return renderWithTheme(<Checkbox {...defaultProps} {...props} />);
};

describe("Checkbox Component", () => {
  const defaultClass =
    "peer h-4 w-4 shrink-0 rounded-sm border border-secondary shadow focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-secondary data-[state=checked]:text-primary-foreground";

  // Snapshot Tests
  it("should match snapshot with default props", () => {
    const { container } = renderComponent();
    expect(container.firstChild).toMatchSnapshot();
  });

  // Basic Rendering
  it("should render checkbox with default props", () => {
    renderComponent();
    const checkbox = screen.getByTestId("checkbox-root");
    expect(checkbox).toBeInTheDocument();
    expect(checkbox).toHaveClass(defaultClass);
  });

  // States
  it("should render in unchecked state by default", () => {
    renderComponent();
    const checkbox = screen.getByTestId("checkbox-root");
    expect(checkbox).toHaveAttribute("data-state", "unchecked");
  });

  it("should render in checked state when checked prop is true", () => {
    renderComponent({ checked: true });
    const checkbox = screen.getByTestId("checkbox-root");
    expect(checkbox).toHaveAttribute("data-state", "checked");
    const indicator = screen.getByTestId("checkbox-indicator");
    expect(indicator).toBeInTheDocument();
    expect(screen.getByTestId("check-icon")).toBeInTheDocument();
  });

  it("should render in disabled state", () => {
    renderComponent({ disabled: true });
    const checkbox = screen.getByTestId("checkbox-root");
    expect(checkbox).toBeDisabled();
  });

  // Event Handlers
  it("should call onCheckedChange when clicked", () => {
    const handleChange = vi.fn();
    renderComponent({ onCheckedChange: handleChange });

    const checkbox = screen.getByTestId("checkbox-root");
    fireEvent.click(checkbox);

    expect(handleChange).toHaveBeenCalledTimes(1);
    expect(handleChange).toHaveBeenCalledWith(true);
  });

  // Custom Styling
  it("should merge custom className with default classes", () => {
    const customClass = "custom-class";
    renderComponent({ className: customClass });
    const checkbox = screen.getByTestId("checkbox-root");
    expect(checkbox).toHaveClass(customClass);
    expect(checkbox).toHaveClass(defaultClass);
  });

  // Accessibility
  it("should have appropriate ARIA attributes", () => {
    const ariaLabel = "Terms and Conditions";
    renderComponent({
      "aria-label": ariaLabel,
      "aria-required": true,
    });

    const checkbox = screen.getByTestId("checkbox-root");
    expect(checkbox).toHaveAttribute("aria-label", ariaLabel);
    expect(checkbox).toHaveAttribute("aria-required", "true");
  });
});
