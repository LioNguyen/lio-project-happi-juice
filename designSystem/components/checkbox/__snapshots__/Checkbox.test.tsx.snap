// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Checkbox Component > should match snapshot with default props 1`] = `
<button
  class="peer h-4 w-4 shrink-0 rounded-sm border border-secondary shadow focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-secondary data-[state=checked]:text-primary-foreground"
  data-state="unchecked"
  data-testid="checkbox-root"
  type="button"
>
  <span
    class="flex items-center justify-center text-current"
    data-testid="checkbox-indicator"
  >
    <div
      data-testid="check-icon"
    />
  </span>
</button>
`;
